components/iam/pkg/oidc/router.go

```diff
@@ -5,16 +5,149 @@
+
+		ctx := r.Context()
+		config, ok := ctx.Value(keyOIDCClientConfig{}).(OIDCClientConfig)
+		if !ok {
+			http.Error(rw, "config not found", http.StatusInternalServerError)
+			return
+		}
+
+		startParams, err := oidcService.GetStartParams(&config)
+		if err != nil {
+			http.Error(rw, "failed to start auth flow", http.StatusInternalServerError)
+			return
+		}
+
+		http.SetCookie(rw, newCallbackCookie(r, nonceCookieName, startParams.Nonce))
+		http.SetCookie(rw, newCallbackCookie(r, stateCookieName, startParams.State))
+
+		http.Redirect(rw, r, startParams.AuthCodeURL, http.StatusTemporaryRedirect)
+	}
+}
+
+func newCallbackCookie(r *http.Request, name string, value string) *http.Cookie {
+	return &http.Cookie{
+		Name:     name,
+		Value:    value,
+		MaxAge:   int(10 * time.Minute.Seconds()),
+		Secure:   r.TLS != nil,
+		SameSite: http.SameSiteLaxMode,
+		HttpOnly: true,
+	}
+}
+
+// The config middleware is responsible to retrieve the client config suitable for request
+func (oidcService *OIDCService) clientConfigMiddleware() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			log.Trace("at config middleware")
+
+			config, err := oidcService.GetClientConfigFromRequest(r)
+			if err != nil {
+				log.Warn("client config not found: " + err.Error())
+				http.Error(rw, "config not found", http.StatusNotFound)
+				return
+			}
+
+			ctx := context.WithValue(r.Context(), keyOIDCClientConfig{}, config)
+			next.ServeHTTP(rw, r.WithContext(ctx))
+		})
+	}
+}
+
+// The OIDC callback handler depends on the state produced in the OAuth2 middleware
+func (oidcService *OIDCService) getCallbackHandler() http.HandlerFunc {
+	return func(rw http.ResponseWriter, r *http.Request) {
+		log.Trace("at callback handler")
+
+		ctx := r.Context()
+		config, ok := ctx.Value(keyOIDCClientConfig{}).(OIDCClientConfig)
+		if !ok {
+			http.Error(rw, "config not found", http.StatusInternalServerError)
+			return
+		}
+		oauth2Result, ok := ctx.Value(keyOAuth2Result{}).(OAuth2Result)
+		if !ok {
+			http.Error(rw, "OIDC precondition failure", http.StatusInternalServerError)
+			return
+		}
+
+		// nonce = number used once
+		nonceCookie, err := r.Cookie(nonceCookieName)
+		if err != nil {
+			http.Error(rw, "nonce not found", http.StatusBadRequest)
+			return
+		}
+
+		result, err := oidcService.Authenticate(ctx, &oauth2Result,
+			config.Issuer, nonceCookie.Value)
+		if err != nil {
+			http.Error(rw, "OIDC authentication failed", http.StatusInternalServerError)
+			return
+		}
+
+		// TODO(at) given the result of OIDC authN, let's proceed with the redirect
+
+		// For testing purposes, let's print out redacted results
+		oauth2Result.OAuth2Token.AccessToken = "*** REDACTED ***"
+
+		var claims map[string]interface{}
+		err = result.IDToken.Claims(&claims)
+		if err != nil {
+			http.Error(rw, err.Error(), http.StatusInternalServerError)
```

Comment:
This should be a user error, no? If the user supplies the wrong claim, we shouldn't return a Server error.

Reply:
[Claims](https://openid.net/specs/openid-connect-core-1_0.html#Claims) is what Gitpod (as the OIDC client) can obtain from the end user. 

Not being able to parse the `id_token` can have these causes:
* unparseable gibberish received from IdP (e.g. their server/lb is misconfigured) 
* otherwise parsing failures => status code 500 it's Gitpod's fault

Also, while the error modes are still relevant, following PRs will remove this section.